mrjob/dataproc.py

```diff
@@ -0,0 +1,873 @@
+
+        Reference...
+        https://cloud.google.com/dataproc/reference/rest/v1/projects.regions.jobs#HadoopJob
+
+        :param step_num:
+        :return:
+        """
+        args = list()
+        file_uris = list()
+        archive_uris = list()
+        properties = dict()
+
+        step = self._get_step(step_num)
+
+        assert step['type'] in ('streaming', 'jar'), 'Bad step type: %r' % (step['type'],)
+
+        # TODO - mtai @ davidmarin - Might be trivial to support jar running, see "mainJarFileUri" of variable "output_hadoop_job" in this function
+        #         https://cloud.google.com/dataproc/reference/rest/v1/projects.regions.jobs#HadoopJob
+        assert step['type'] == 'streaming', 'Jar not implemented'
+        main_jar_uri = _HADOOP_STREAMING_JAR_URI
+
+
+        # TODO - mtai @ davidmarin - Not clear if we should move _upload_args to file_uris, currently works fine as-is
+        args.extend(self._upload_args(self._upload_mgr))
+
+        # TODO - mtai @ davidmarin - Not clear if we should move some of these to properties dict to file_uris, currently works fine as-is
+        args.extend(self._hadoop_args_for_step(step_num))
+
+        mapper, combiner, reducer = (self._hadoop_streaming_commands(step_num))
+
+        if mapper:
+            args += ['-mapper', mapper]
+
+        if combiner:
+            args += ['-combiner', combiner]
+
+        if reducer:
+            args += ['-reducer', reducer]
+
+        for current_input_uri in self._step_input_uris(step_num):
+            args += ['-input', current_input_uri]
+
+        args += ['-output', self._step_output_uri(step_num)]
+
+        # TODO - mtai @ davidmarin - Add back support to specify a different jar URI
+        output_hadoop_job = {
+            "args": args,
+            "fileUris": file_uris,
+            "archiveUris": archive_uris,
+            "properties": properties,
+            "mainJarFileUri": main_jar_uri
+        }
+        return output_hadoop_job
+
+    def _launch_dataproc_cluster(self):
+        """Create an empty cluster on Dataproc, and set self._gcp_cluster to
+        its ID."""
+        self._wait_for_gcs_eventual_consistency()
+
+        self._gcp_cluster_start = time.time()
+
+        # make sure we can see the files we copied to GCS
+        log.info('Creating Dataproc Hadoop cluster')
+
+        # NOTE - mtai @ davidmarin - We create the cluster if its missing, otherwise we join an existing one
+        # NOTE - mtai @ davidmarin - The cluster MUST be created before submitting jobs, otherwise job submission fails
+
+        # TODO - mtai @ davidmarin - change the way we auto-create cluster names
+        # "clusterName must be a match of regex '(?:[a-z](?:[-a-z0-9]{0,53}[a-z0-9])?).'"
+        self._gcp_cluster = self._gcp_cluster or "default-%s" % self._opts['owner']
+
+        try:
+            self._get_cluster()
+            log.info('Adding our job to existing cluster %s' % self._gcp_cluster)
```

Comment:
A couple of things to think about that might inform the design of the Dataproc runner:

Dataproc lets you run multiple jobs on the same cluster; when is this actually useful? (We already have a cloud, so we don't need Hadoop/YARN on a cluster to act as a cloud within a cloud). Some ideas:
 * Latency, maybe? Like, I have a very short job, and I'd rather have it compete with other jobs on an existing cluster than wait for a new cluster to spin up?

In theory, as long as step output goes to GCS and not HDFS, a job could, in theory, run every step on a different cluster, shopping around on a step-by-step basis. Honestly, I can't see much advantage to this; unlike on EMR (where we never tried this), you're going to be paying by the minute for any cluster you use, so there aren't really "available" clusters just waiting around for you to use their idle time.